Automattic/wp-calypso
# [ETK/error-reporting] Activate sentry for a segment of users

## Changes proposed in this Pull Request

Load Sentry in the error-reporting ETK module that applies to wp-admin + Gutenberg, but only activate it for a segment of users (1% for now, using a simple modulus calc), and a12s' blogs that have the `error-reporting-use-sentry` sticker* applied. Keep activating our homebrew error-reporter for all other users.

*_Add the sticker with `wp --url=<your-wpcom-test-site-url> blog-stickers add --sticker=error-reporting-use-sentry --who=<your-wpcom-username>`. To remove just swap the `add` command with `remove`._

The goal is to experiment with Sentry and test its feasibility, while leaving our homebrew error-reporting active for all others. We're still using the same ETK module to do so (as it can make it easier to ship it to WoA later).

If everything goes well, we might deprecate our homebrew solution in favor of Sentry. More details: p9oQ9f-Ay-p2#comment-1075 and p9oQ9f-zY-p2#comment-1055 under the `Next` section 🤞🏻 

## Testing instructions

⚠️ Before testing any of the following scenarios, be sure to activate jsconcat in your WPCOM sandbox (apply this diff: D63665-code). This will start serving files from CDN domains (**i.e `s0.wp.com` -- be sure to sandbox this, too!**) which gets us closer to a production environment and allows us to test for CORS issues (for more details, see this systems-request: pMz3w-cCq-p2) ⚠️ 

### Happy-path: Sentry works when activated, verify it correctly captures JS errors on the editor page

1. Apply the `error-reporting-user-sentry` sticker to your sandboxed test blog (as an a11n, you can apply the `error-reporting-use-sentry` sticker to a blog (see command above) to instruct this module to activate Sentry);
2. Checkout this branch, `cd` to `apps/editing-toolkit`, and run `yarn dev --sync` to build and sync the custom ETK build to your sandbox;
2. You can then trigger an error manually [like this](https://github.com/Automattic/wp-calypso/pull/49485#issuecomment-828029870) from the editor page. Be sure to select the wp-admin iframe context (Gutenframe) if you're running the site through Calypso, or the error will be reported to the Calypso error reporter, instead!
3. Login to the A8C Sentry account and verify that the error has been captured in the `wpcom-gutenberg-wp-admin` project. Follow the instructions here: PCYsg-ldt-p2if you're not sure how to create your account and login).
4. Test "head" errors. These are errors that might happen before the error reporter (in this case, Sentry) has been loaded. In this case, it should be caught by and stored in a window-wise array, and then once the error reporting module is loaded - and after it activates Sentry  - it will then iterate through all of them and send them over the wire. There's currently some test code that will generate some exceptions for you. Verify that Sentry has 3 recent `Head error ${count} from the top-level document` errors shortly after you loaded the page.

### Additional scenarios

---

**UPDATE**: I [added](https://github.com/Automattic/wp-calypso/pull/54257/commits/05664a223a2d6bb04d4e2bb9044344373cc35f62) unit tests for the modulus segmentation/activation logic. See [here](https://github.com/Automattic/wp-calypso/pull/54257/commits/05664a223a2d6bb04d4e2bb9044344373cc35f62). I encourage to test these manually, too, since these are not integration tests, but we already have some further confidence in the core activation logic.

To run the tests:

`apps/editing-toolkit (try/sentry-for-a-segment-of-users●)$ yarn test:php`

@noahtallen Gave a good primer in this thread (slack): p1620858708150500-slack-C7YPUHBB2.

---

#### When loading a blog without the `error-reporting-use-sentry` applied as an a11n, the homebrew error is activated
1. Make sure your blog doesn't have the aforementioned sticker, remove it if necessary;
2. Checkout this branch, `cd` to `apps/editing-toolkit`, and run `yarn dev --sync` to build and sync the custom ETK build to your sandbox;
3. Load the editor and you should see `Activating homebrew error reporting!`, which means our version of error-reporting has been loaded instead _(`homebrew` here has nothing to do with the packaging tool, but is a way to refer to our own error-reporting solution, I'm trying to come up with a better term :))_

#### When loading a blog with the `error-reporting-use-sentry` applied as an a11n, Sentry is activated
1. Apply the `error-reporting-user-sentry` to your sandboxed test blog;
2. Checkout this branch, `cd` to `apps/editing-toolkit`, and run `yarn dev --sync` to build and sync the custom ETK build to your sandbox;
3. Load the editor and you should see `Activating Sentry!`
4. You can refer to the happy-path scenario above to further test Sentry and verify that it's capturing the errors.

#### Sentry is not activated for non-a12s:
1. In your sandbox, modify the `is_automattician` function to always return `false` (search for `function is_automattician` in opengrok to find where it's defined)
2. Checkout this branch, `cd` to `apps/editing-toolkit`, and run `yarn dev --sync` to build and sync the custom ETK build to your sandbox;
3. Load the editor and you should see `Activating homebrew error reporting!`, which means our version of error-reporting has been loaded instead.

#### Test the modulus segmentation logic works and that it loads the correct error reporter for each segment
1. Find out your user_id, you can output it using l( get_current_user_Id ) somewhere in the index.php for the error-reporting ETK module;
2. Let’s say your user_id is 7922883, this means you’d fall into the 84% segment. The default value in the test PR is `10`, so it should not load Sentry for you by default (it'd only load ids ending in 0-9). Try changing the `$current_segment` variable's value [here](https://github.com/Automattic/wp-calypso/blob/try/sentry-for-a-segment-of-users/apps/editing-toolkit/editing-toolkit-plugin/error-reporting/index.php#L54) to `84`, and then you should see `Activating Sentry` in the console.
3. Then change it to < 84 and you should see `Activating homebrew error reporting!` instead.

## Todo

- [ ] ~~Move the Sentry account data over to the config, in order to have different environments (staging, prod) - for now we'd probably only activate it on prod, though. See discussion: p1626280547282100/1625249660.115200-slack-C7YPUHBB2.~~

## Outstanding questions
- [ ] ~~Should we mutually activate each error reporting solution or would it make sense to activate both Sentry and our our error reporting at the same time?~~
- [x] What Sentry account to use? We can use the one I created using my A8C email (search for `Test Sentry account for WPCOM`  in the secret store) and move it to a Sentry-specific mailbox. This would require a systemsrequest to ask for the mailbox to be created.
